### PR TITLE
Interactive atom support

### PR DESCRIPTION
**Pull request by wpf500**
## What does this change?

Render interactive atoms on article and interactive content
## What is the value of this and can you measure success?

Interactive content becomes a first-class citizen. HTML baked into the page from the start, faster loading, resource sharing (i.e. fonts) even in iframes.
## Does this affect other platforms - Amp, Apps, etc?

Yes, separate work going on in apps. Off-platform will display the default HTML from CAPI
## Screenshots

The faster one is an atom
ezgif com-resize

The faster one is an atom
ezgif com-video-to-gif

No JavaScript, the one that loads is an atom
image

Unique, immersive, beautiful content. The faster one is an atom
ezgif com-video-to-gif 1
## Request for comment

Not quite finished yet but putting it up to get the review process started.

I suspect there's going to be some comments about the iframe code, I'm just leaving it there to start a discussion. The reason I'm not just using https://github.com/guardian/fence for iframes is because it doesn't support the content changing height (e.g. on resize). I also think the iframe should resize itself as soon as possible rather than waiting for the standard JS to kick in.

I've also removed some no-JS fallback styling for interactives (added in https://github.com/guardian/frontend/pull/10545), it doesn't make current no-JS look any better, and ruins the atoms
### TODO:
- [x] Font sharing
- [x] `srcdoc` fallback (with fence?)

### CODE REVIEW

**Comment by wpf500** [comment]
@GHaberis @gtrufitt @stephanfowler @sammorrisdesign

**Comment by wpf500** [comment]
This still has support for https://github.com/guardian/frontend/pull/14402, but I anticipate removing it. Not sure if there will be any interactives that need to be retrofitted @sammorrisdesign?

**Comment by sammorrisdesign** [comment]
@wpf500 Very happy to remove the `srcdoc` support. I retrofitted one article to work in that way but will happily use it as a test case for this instead.

**Comment by paperboyo** [comment]
💎! I’m interested in fonts sharing, esp. in apps...

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (wpf500) [inline common/app/views/fragments/atoms/interactive.scala.html]: Before anyone says it, this is obviously not very nice. But I'm interested to talk about how we can get fonts loaded into the iframes fast.

Comment (gtrufitt) [inline common/app/views/fragments/atoms/interactive.scala.html]: Maybe @regiskuckaertz will have an opinion on this?

Comment (regiskuckaertz) [inline common/app/views/fragments/atoms/interactive.scala.html]: (although I would move that fragment in the `HEAD` to avoid a complete reflow)

File: common/app/model/content/Atoms.scala
Comment (TBonnin) [inline common/app/model/content/Atoms.scala]: Would it make sense to use `partialFunction` here to avoid the `asInstanceOf` and `try/catch` dance here? Something like that:
```
    def make(content: contentapi.Content): Option[Atoms] = {

    def extractAtoms[T](atoms: Option[Seq[AtomApiAtom]], dataToAtom: PartialFunction[AtomData, T]): Seq[T] = atoms.fold(Seq.empty[T])(_.map { atom => dataToAtom(atom.data)})

    val atomDataToQuiz: PartialFunction[AtomData, Quiz] = { case data: AtomData.Quiz => Quiz.make(content.id, data.quiz) }
    val atomDataToMedia: PartialFunction[AtomData, MediaAtom] = { case data: AtomData.Media => MediaAtom.mediaAtomMake(content.id, data.media) }
    val atomDataToInteractive: PartialFunction[AtomData, InteractiveAtom] = { case data: AtomData.Interactive => InteractiveAtom.make(content.id, data.interactive) }

    content.atoms.map { atoms =>
      Atoms(
        quizzes = extractAtoms(atoms.quizzes, atomDataToQuiz),
        media = extractAtoms(atoms.media, atomDataToMedia),
        interactives = extractAtoms(atoms.quizzes, atomDataToInteractive)
      )
    }
  }
```

File: applications/app/views/fragments/interactiveBody.scala.html
Comment (gtrufitt) [inline applications/app/views/fragments/interactiveBody.scala.html]: Is this a question that needs answering❓

**Review by TBonnin — CHANGES_REQUESTED** [review]

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (gtrufitt) [inline common/app/views/fragments/atoms/interactive.scala.html]: Can we abstract these scripts out into an interactives folder in JS files and use the `@InlineJs` helper. That way we can still integrate it into the build.

File: common/app/model/content/Atoms.scala
Comment (wpf500) [inline common/app/model/content/Atoms.scala]: Happy to do so, but should I change the other atom `make` methods to `apply` too for consistency?

File: applications/app/views/fragments/interactiveBody.scala.html
Comment (wpf500) [inline applications/app/views/fragments/interactiveBody.scala.html]: I tried to track down if this was ever used. I can't see any scenario where an interactive made now would have `maybeBody == None` so was hoping to clean it up. But yeah, still a question!

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (wpf500) [inline common/app/views/fragments/atoms/interactive.scala.html]: Not sure I understand this question, but about to change this bit anyway

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (wpf500) [inline common/app/views/fragments/atoms/interactive.scala.html]: Will do

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (gtrufitt) [inline common/app/views/fragments/atoms/interactive.scala.html]: So, we're not fencing here, right? My concern would be that `srcdoc` isn't that well supported:

image

So what is your plan for handling IE11, Edge if not using fence.js? I'm surprised it didn't handle resizing of iframes by content, I don't remember interactive embeds always having scroll bars?

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (gtrufitt) [inline common/app/views/fragments/atoms/interactive.scala.html]: Yeah, just realised that iframeBody inherits body, nvm

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (wpf500) [inline common/app/views/fragments/atoms/interactive.scala.html]: `srcdoc` fallback still to come, see the description.

File: common/app/model/content/Atoms.scala
Comment (TBonnin) [inline common/app/model/content/Atoms.scala]: If possible I would say yes

File: common/app/views/fragments/atoms/interactive.scala.html
Comment (gtrufitt) [inline common/app/views/fragments/atoms/interactive.scala.html]: 🆗

Comment (wpf500) [inline common/app/views/fragments/atoms/interactive.scala.html]: The main reason I haven't done this yet is because I'm in two minds about the right solution. My first thought was just to edit fence to support resizing, but fence still has the 2-3 sec execution delay that most JS has on frontend.

Personally, I like that the iframes now resize instantly (less page janking), but I'm not sure if it goes against performance guidelines or something. If it doesn't then I'd probably just implement a simpler `srcdoc` fallback (without all the resizing that `fence` does).

I don't know very much about the implications here in terms of performance vs user experience, so would be great to chat in person.

File: common/app/model/content/Atoms.scala
Comment (wpf500) [inline common/app/model/content/Atoms.scala]: @TBonnin, unfortunately I've tried it and it messes with Play JSON's implicit conversion macro, which IIRC uses the `apply` signature to create the `Json.Format[T]`. This is probably why the other atoms used `make`

File: common/app/templates/inlineJS/nonBlocking/interactiveResize.scala.js
Comment (regiskuckaertz) [inline common/app/templates/inlineJS/nonBlocking/interactiveResize.scala.js]: Won't this trigger an infinite resize loop?

Comment (gtrufitt) [inline common/app/templates/inlineJS/nonBlocking/interactiveResize.scala.js]: `resize()` *then* add the listener?

File: common/app/model/content/Atoms.scala
Comment (TBonnin) [inline common/app/model/content/Atoms.scala]: What about passing the ApiAtom and matching using `isInstanceOf` as well as using `collect` rather than `map`

Like this
```
def make(content: contentapi.Content): Option[Atoms] = {

    val apiAtomToQuiz: PartialFunction[AtomApiAtom, Quiz] = {
      case atom if atom.data.isInstanceOf[AtomData.Quiz] => Quiz.make(content.id, atom.data.asInstanceOf[AtomData.Quiz].quiz)
    }
    val apiAtomToMedia: PartialFunction[AtomApiAtom, MediaAtom] = {
      case atom if atom.data.isInstanceOf[AtomData.Media] => MediaAtom.mediaAtomMake(content.id, atom.data.asInstanceOf[AtomData.Media].media)
    }
    val apiAtomToInteractive: PartialFunction[AtomApiAtom, InteractiveAtom] = {
      case atom if atom.data.isInstanceOf[AtomData.Interactive] => InteractiveAtom.make(atom.id, atom.data.asInstanceOf[AtomData.Interactive].interactive)
    }

    def extractAtoms[T](atoms: Option[Seq[AtomApiAtom]], apiAtomtoAtom: PartialFunction[AtomApiAtom, T]): Seq[T] = atoms.fold(Seq.empty[T])(_.collect(apiAtomtoAtom))

    content.atoms.map { atoms =>
      Atoms(
        quizzes = extractAtoms(atoms.quizzes, apiAtomToQuiz),
        media = extractAtoms(atoms.media, apiAtomToMedia),
        interactives = extractAtoms(atoms.quizzes, apiAtomToInteractive)
      )
    }
  }
```

File: common/app/templates/inlineJS/nonBlocking/interactiveFonts.scala.js
Comment (gtrufitt) [inline common/app/templates/inlineJS/nonBlocking/interactiveFonts.scala.js]: Might wanna wrap this and resize in `iife` to avoid global scope? Maybe `try catch` to harden it.

**Review by TBonnin — APPROVED** [review]

**Review by gtrufitt — APPROVED** [review]
Minor changes with the `iife` but LGTM 👍 

I'd wait until the bright light of day to merge though ;)

Fantastic work!

File: common/app/templates/inlineJS/nonBlocking/interactiveFonts.scala.js
Comment (wpf500) [inline common/app/templates/inlineJS/nonBlocking/interactiveFonts.scala.js]: I'm going to leave this for now. The `resize` method especially could actually be useful in the embeds for which want to update their size

**Pull request merged**
